{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before anything: Start your mySQL Server from terminal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Open both files en reading mode\n",
    "\n",
    "germfile = open(\"Germplasm.tsv\", \"r\")\n",
    "genefile = open(\"LocusGene.tsv\", \"r\")\n",
    "#print(germfile.read()) #If I want, I can check that files have been opened correctly\n",
    "#print(genefile.read())\n",
    "germfile.seek(0) # I want to be sure that my files are read since the beginning\n",
    "genefile.seek(0)\n",
    "\n",
    "# 2. Use regular expressions for finding all the times that AGI Locus Codes appear in both files by 're.findall'\n",
    "#I am going to generate two lists with AGI Codes in the same appearance order that in tsv files\n",
    "\n",
    "import re  # this brings the python regular expression object into your program\n",
    "match_germ = re.findall( r'AT\\dG\\d{5}', germfile.read(), re.I) #re.I makes re.findall case insensitive\n",
    "match_gene = re.findall( r'AT\\dG\\d{5}', genefile.read(), re.I)\n",
    "#print(match_germ)\n",
    "#print(match_gene)\n",
    "\n",
    "# 3. Check that both files have AGI Codes in the same lines\n",
    "\n",
    "if match_germ == match_gene:\n",
    "    print(\"All our data is in the same lines!! We are very tidy!!\")\n",
    "    \n",
    "else: \n",
    "    print(\"Sorry! Our data is a disaster...\")\n",
    "\n",
    "# 4. Close files\n",
    "\n",
    "germfile.close()\n",
    "genefile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. I create a new database by sqlMagic and I check that it exists.\n",
    "\n",
    "%sql create database exam2\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. I create 1st table by sqlMagic and I check its features.\n",
    "%sql use exam2 #I have to select the database\n",
    "%sql drop table germplasm\n",
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(20) NOT NULL, germplasm VARCHAR(50) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed VARCHAR(20) NOT NULL);\n",
    "%sql DESCRIBE germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3 I create 2nd table by sqlMagic and I check its features.\n",
    "#%sql use exam2 #I have to select the database\n",
    "%sql drop table gene\n",
    "%sql CREATE TABLE gene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(20) NOT NULL, Gene VARCHAR(15) NOT NULL, ProteinLength VARCHAR(20) NOT NULL);\n",
    "%sql DESCRIBE gene\n",
    "#Both tables are connected by the 'Locus' field. This guarantees the 1:1 relationship."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. I add information to the table 'gene' by pymysql from LocusGene.tsv\n",
    "\n",
    "import csv # I have to import specific package into the program for reading csv files\n",
    "\n",
    "import pymysql.cursors # I have to import specific package into the program for using pymysql\n",
    "\n",
    "connection = pymysql.connect(host='localhost', #With this information I basically provide to pymysql the location of my database.\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False # As I am going to commit manually, I put this on False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open(\"LocusGene.tsv\") as tsvfile:\n",
    "            geneinfo = csv.DictReader(tsvfile, delimiter=\"\\t\") \n",
    "            #'DictReader has special features for reading files: 'delimiter' indicates the separator character between fields \\\n",
    "            # In this case, the separator character is 'tab'\n",
    "            for row in geneinfo:\n",
    "                sql = \"\"\"INSERT INTO gene (Locus, Gene, ProteinLength) \\\n",
    "                VALUES ('{}', '{}', '{}')\"\"\".format(row[\"Locus\"],row[\"Gene\"],row[\"ProteinLength\"])\n",
    "                #this command inserts the different data of each header field in the table and makes it row-to-row.\n",
    "                cursor.execute(sql) #this command execute the previous order\n",
    "\n",
    "    connection.commit() # I commit manually\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close() #I close pymysql. This is necessary for its good running\n",
    "\n",
    "%sql SELECT * FROM gene; # I check the table after the insertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. I add information to the table 'germplasm' by pymysql from Germplasm.tsv\n",
    "\n",
    "import csv # I have to import specific package into the program for reading csv files\n",
    "\n",
    "import pymysql.cursors # I have to import specific package into the program for using pymysql\n",
    "\n",
    "connection = pymysql.connect(host='localhost', #With this information I basically provide to pymysql the location of my database.\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit = False # As I am going to commit manually, I put this on False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open(\"Germplasm.tsv\") as tsvfile:\n",
    "            geneinfo = csv.DictReader(tsvfile, delimiter=\"\\t\") \n",
    "            #'DictReader has special features for reading: 'delimiter' indicates the separator character between fields \\ \n",
    "            # In this case is 'tab'\n",
    "            for row in geneinfo:\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed) \\\n",
    "                VALUES ('{}', '{}', '{}', '{}')\"\"\".format(row[\"Locus\"],row[\"germplasm\"],row[\"phenotype\"],row[\"pubmed\"])\n",
    "                #this command inserts the different data of each field in the table and makes it row-to-row.\n",
    "                cursor.execute(sql) #this command execute the previous order\n",
    "\n",
    "    connection.commit() # I commit manually\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close() #I close pymysql. This is necessary.\n",
    "\n",
    "%sql SELECT * FROM germplasm; # I check the table after the insertion"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.1**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. I generate a unique table with all the data and I link the data to a variable.\n",
    "\n",
    "mother_table = %sql SELECT * FROM gene, germplasm WHERE gene.Locus = germplasm.Locus\n",
    "\n",
    "# 2. I write the clean content of the generated variable in a csv file.\n",
    "\n",
    "import csv\n",
    "import io\n",
    "\n",
    "with open('report_exam2.csv', 'w', newline='') as csvfile:\n",
    "    reportwriter = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "    reportwriter.writerow([\"1st Report\"]) # I write the title of the report\n",
    "    reportwriter.writerow((\"Locus\", \"Gene\", \"ProteinLength\", \"Germplasm\", \"Phenotype\", \"Pubmed\"))  # This line writes the header\n",
    "    for row in mother_table: #This command line iterates along the table for filling all the rows\n",
    "        reportwriter.writerow([row[\"Locus\"],row[\"Gene\"],row[\"ProteinLength\"],row[\"germplasm\"],row[\"phenotype\"],row[\"pubmed\"]])\n",
    "csvfile.close\n",
    "\n",
    "#3. I check the content of the report. I think that the operation has been completed succesfully!\n",
    "csvfile = io.open(\"report_exam2.csv\")\n",
    "print(csvfile.read())\n",
    "csvfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.2**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. I generate a unique table with all the data and I link the data to a variable.\n",
    "\n",
    "mother_table = %sql SELECT * FROM gene, germplasm WHERE gene.Locus = germplasm.Locus\n",
    "\n",
    "# 2. I append the required content in the previous csv file.\n",
    "\n",
    "import csv\n",
    "import io\n",
    "\n",
    "with open('report_exam2.csv', 'a', newline='') as csvfile: #The'a' argument indicates that the written information will be appended.\n",
    "    reportwriter = csv.writer(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "    reportwriter.writerow([\"\"]) # I write an 'empty' line\n",
    "    reportwriter.writerow([\"2nd Report\"])\n",
    "    reportwriter.writerow((\"Locus\", \"Gene\", \"ProteinLength\", \"Germplasm\", \"Phenotype\", \"Pubmed\"))  # This line writes the header\n",
    "    for row in mother_table: #This command line iterates along all the rows of the table\n",
    "        if row[\"Gene\"] == \"SKOR\" or row[\"Gene\"] == \"MAA3\": # This clause makes that only SKOR and MAA3 files can be added\n",
    "            reportwriter.writerow([row[\"Locus\"],row[\"Gene\"],row[\"ProteinLength\"],row[\"germplasm\"],row[\"phenotype\"],row[\"pubmed\"]])\n",
    "        else:\n",
    "            reportwriter.writerow([\"This is TOP SECRET information\"]) # In this case, I have wanted to play a little with the else clause \\\n",
    "            # However, this is not necessary for the good running of the if clause.\n",
    "            \n",
    "csvfile.close\n",
    "\n",
    "# 3. I check the content of the report. I think that the operation has been completed succesfully!\n",
    "csvfile = io.open(\"report_exam2.csv\")\n",
    "print(csvfile.read())\n",
    "csvfile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.3**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. I open LocusGene.tsv file in reading mode.\n",
    "\n",
    "genefile = open(\"LocusGene.tsv\", \"r\")\n",
    "\n",
    "# 2. I use regular expressions for finding all the times that AGI Locus Codes appear in LocusGene.tsv by're.findall'. \\\n",
    "# Besides, I am going to generate five variables, one for each chromosome.\n",
    "\n",
    "import re  # this brings the python regular expression object into your program \n",
    "\n",
    "genefile.seek(0) # I want to be sure that my files are read since the beginning \n",
    "\n",
    "match_chr1_genes = re.findall( r'AT1G\\d{5}', genefile.read(), re.I) # The third character indicates the number of the chromosome. This character is the unique thing that changes in each \"harvest\". \n",
    "#re.findall generates a list that is associated to the variable match_chr1_genes\n",
    "chr1_counts = len(match_chr1_genes) # Since match_chr1_genes is a list, I can measure the length of the list and link it to variable chr1_counts.\n",
    "\n",
    "genefile.seek(0) #I have to add this line before each \"harvest\". If not, harvest will begin since the bottom of the file and harvest will be vain\n",
    "\n",
    "match_chr2_genes = re.findall( r'AT2G\\d{5}', genefile.read(), re.I) #re.I makes re.findall case insensitive\n",
    "chr2_counts = len(match_chr2_genes)\n",
    "\n",
    "genefile.seek(0)\n",
    "\n",
    "match_chr3_genes = re.findall( r'AT3G\\d{5}', genefile.read(), re.I) #re.I makes re.findall case insensitive\n",
    "chr3_counts = len(match_chr3_genes)\n",
    "\n",
    "genefile.seek(0)\n",
    "\n",
    "match_chr4_genes = re.findall( r'AT4G\\d{5}', genefile.read(), re.I) #re.I makes re.findall case insensitive\n",
    "chr4_counts = len(match_chr4_genes)\n",
    "\n",
    "genefile.seek(0)\n",
    "\n",
    "match_chr5_genes = re.findall( r'AT5G\\d{5}', genefile.read(), re.I) #re.I makes re.findall case insensitive\n",
    "chr5_counts = len(match_chr5_genes)\n",
    "\n",
    "# 3. I append to the file report.exam2.csv the 3rd report with the information of the five previous variables\n",
    "\n",
    "report = open(\"report_exam2.csv\", \"a\")\n",
    "report.write(\"\\n3rd Report\\nThe 1st chromosome has \" + str(chr1_counts) + \" genes\" \\\n",
    "             + \".\\nThe 2nd chromosome has \" + str(chr2_counts) + \" genes\" \\\n",
    "             + \".\\nThe 3rd chromosome has \" + str(chr3_counts) + \" genes\" \\\n",
    "             + \".\\nThe 4th chromosome has \" + str(chr4_counts) + \" genes\" \\\n",
    "             + \".\\nThe 5th chromosome has \" + str(chr5_counts) + \" genes\")\n",
    "report.close()\n",
    "\n",
    "#4. I check the content of the file for look at the correct appending.\n",
    "\n",
    "checkcontent = open(\"report_exam2.csv\", \"r\")\n",
    "print(checkcontent.read())\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Appendix to Problem 4.3\n",
    "\n",
    "#I know that the 2nd step of the previous code is repeating. This can be solved writing a function, like:\n",
    "\n",
    "def gene_counts(chr_number):\n",
    "    genefile.seek(0)\n",
    "    match_chr_genes = re.findall(r'AT'chr_number'G\\d{5}', genefile.read(), re.I) #re.I makes re.findall case insensitive\n",
    "    chr_counts = len(match_chr_genes)\n",
    "    return chr_counts\n",
    "\n",
    "#With this function, I could write the report of this way:\n",
    "\n",
    "report = open(\"report_exam2.csv\", \"a\")\n",
    "report.write(\"\\n3rd Report\\nThe 1st chromosome has \" + str(gene_counts(1)) + \" genes\" \\\n",
    "             + \".\\nThe 2nd chromosome has \" + str(gene_counts(2)) + \" genes\" \\\n",
    "             + \".\\nThe 3rd chromosome has \" + str(gene_counts(3)) + \" genes\" \\\n",
    "             + \".\\nThe 4th chromosome has \" + str(gene_counts(4)) + \" genes\" \\\n",
    "             + \".\\nThe 5th chromosome has \" + str(gene_counts(5)) + \" genes\")\n",
    "report.close()\n",
    "\n",
    "#This strategy could be much more friendly and efficient. However, the problem is that it doesn't work. \n",
    "#I would like to know if it is possible to add a variable to a regular expression, and how it is possible."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Problem 4.4**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. I will obtain the data (the protein sizes) from LocusGene.tsv.\n",
    "\n",
    "import csv\n",
    "import re\n",
    "\n",
    "with open('LocusGene.tsv') as csvfile: #I open the file\n",
    "    gene_info = list(csv.DictReader(csvfile, delimiter=\"\\t\")) # I transform the info of the tsv file in a list\n",
    "    sum_prot_chr1 = 0 # I initialize all the variables that I am going to iterate. This is important if you run the code more of one times.\n",
    "    sum_prot_chr2 = 0\n",
    "    sum_prot_chr3 = 0\n",
    "    sum_prot_chr4 = 0\n",
    "    sum_prot_chr5 = 0\n",
    "    for row in gene_info: # I iterate for each row of the table (that has been transformed into a list)\n",
    "        for AGI_Code in match_chr1_genes: # I iterate for each one of my five lists of genes/chromosome (generated in the previous problem)\n",
    "            if row['Locus'] == AGI_Code:\n",
    "                sum_prot_chr1 += int(row['ProteinLength']) #For each match, I sum the size of the protein associated to the match\n",
    "                #Finally, I obtain the sum of all the proteins associated to the 1st chromosome\n",
    "        for AGI_Code in match_chr2_genes:\n",
    "            if row['Locus'] == AGI_Code:\n",
    "                sum_prot_chr2 += int(row['ProteinLength'])\n",
    "        for AGI_Code in match_chr3_genes:\n",
    "            if row['Locus'] == AGI_Code:\n",
    "                sum_prot_chr3 += int(row['ProteinLength'])\n",
    "        for AGI_Code in match_chr4_genes:\n",
    "            if row['Locus'] == AGI_Code:\n",
    "                sum_prot_chr4 += int(row['ProteinLength'])\n",
    "        for AGI_Code in match_chr5_genes:\n",
    "            if row['Locus'] == AGI_Code:\n",
    "                sum_prot_chr5 += int(row['ProteinLength'])\n",
    "    avg_prot_chr1 = sum_prot_chr1 / len(match_chr1_genes) # I divide the total sum of aminoacids associated to each one \\\n",
    "    # of the 5 chromosomes between the total number of genes belonging to each chromosome\n",
    "    avg_prot_chr2 = sum_prot_chr2 / len(match_chr2_genes)\n",
    "    avg_prot_chr3 = sum_prot_chr3 / len(match_chr3_genes)\n",
    "    avg_prot_chr4 = sum_prot_chr4 / len(match_chr4_genes)\n",
    "    avg_prot_chr5 = sum_prot_chr5 / len(match_chr5_genes)\n",
    "\n",
    "# 2. I write the report in report_exam2 with the results.\n",
    "\n",
    "report = open(\"report_exam2.csv\", \"a\")\n",
    "report.write(\"\\n\\n4st Report\\nThe average lenght of the proteins of the 1st chromosome is of \" + str(avg_prot_chr1) + \" aminoacids\" \\\n",
    "             + \".\\nThe average lenght of the proteins of the 2nd chromosome is of \" + str(avg_prot_chr2) + \" aminoacids\" \\\n",
    "             + \".\\nThe average lenght of the proteins of the 2nd chromosome is of \" + str(avg_prot_chr3) + \" aminoacids\" \\\n",
    "             + \".\\nThe average lenght of the proteins of the 2nd chromosome is of \" + str(avg_prot_chr4) + \" aminoacids\" \\\n",
    "             + \".\\nThe average lenght of the proteins of the 2nd chromosome is of \" + str(avg_prot_chr5) + \" aminoacids\")\n",
    "report.close()\n",
    "    \n",
    "# 3. I check the content of the file for look at the correct appending.\n",
    "\n",
    "checkcontent = open(\"report_exam2.csv\", \"r\")\n",
    "print(checkcontent.read())\n",
    "checkcontent.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
